METAPROGRAMMING.md

```diff
@@ -1,86 +1,266 @@
+    return value.bind(this);
+  }
   delete descriptor.value;
-  return { kind: "field", key, placement, descriptor, initializer };
+  return { kind: 'field', key, placement, descriptor, initializer };
 }
 
 // Whenever a read or write is done to a field, call the render()
 // method afterwards. Implement this by replacing the field with
 // a getter/setter pair.
-function observed({kind, key, placement, descriptor, initializer}) {
-  assert(kind == "field");
-  assert(placement == "own");
+function observed({ kind, key, placement, descriptor, initializer }) {
+  assert(kind == 'field');
+  assert(placement == 'own');
   // Create a new anonymous private name as a key for a class element
   let storage = decorators.PrivateName();
-  let underlyingDescriptor = { enumerable: false, configurable: false, writable: true };
-  let underlying = { kind, key: storage, placement, underlyingDescriptor, initializer };
+  let underlyingDescriptor = {
+    enumerable: false,
+    configurable: false,
+    writable: true
+  };
+  let underlying = {
+    kind,
+    key: storage,
+    placement,
+    underlyingDescriptor,
+    initializer
+  };
   return {
-    kind: "method",
+    kind: 'method',
     key,
     placement,
     descriptor: {
-      get() { storage.get(this); },
+      get() {
+        storage.get(this);
+      },
       set(value) {
         storage.set(this, value);
         // Assume the @bound decorator was used on render
         window.requestAnimationFrame(this.render);
       },
       enumerable: descriptor.enumerable,
-      configurable: descriptor.configurable,
+      configurable: descriptor.configurable
     },
     extras: [underlying]
   };
-  
 }
 ```
+
+## Class element finisher
+
+Here is an example that uses the `finisher` property for a class element:
+
+```js
+class User {
+  @readonly id;
+  email;
+
+  constructor(id = null, email = null) {
+    this.id = id;
+    this.email = email;
+  }
+}
+
+function readonly(elementDecriptor) {
+  // we don't want to set `writable` to false until after the property has been initialized
+  elementDecriptor.finisher = () => {
+    elementDecriptor.descriptor.writable = false;
+  };
+  return elementDecriptor;
+}
+```
+
+The above code is functionally equivalent to the following:
+
+```js
+class User {
+  id;
+  email;
+
+  constructor(id = null, email = null) {
+    this.id = id;
+    this.email = email;
+    // finisher
+    Object.defineProperty(this, 'id', {
+      writable: false,
+      // configurable: false and enumerable: true are the default for public class fields
+      configurable: false,
+      enumerable: true
+    });
+  }
+}
+```
+
+<a id="bound-decorator-notes"></a>
+## Notes about the `@bound` decorator
+
+You may have seen the following pattern for creating bound methods and be wondering why a decorator would be preferred for this:
+
+```js
+class Counter extends HTMLElement {
+  @observed #x = 0;
+
+  constructor() {
+    super();
+    this.onclick = this.clicked;
```

Comment:
I literally just took the example from the readme and changed `clicked` from a regular method to an arrow function, but now that you mention it I agree this is confusing. I could change it so that the arrow function is `onclick` and there's no `clicked` function anymore, but that is a bit of an issue for my section on inheritance, where it would be a little weird to talk about calling `onclick` directly rather than the browser doing it. Maybe the inheritance section should use a different example, hmm...

Reply:
Personally I think you should have all three of the examples I mentioned; for each one, talking about its pros and cons, and then talking about the last example (instance method + bound decorator) and why it's the best choice.

Reply:
For the sake of comparison/explanation, it would be nice to have a simple example just using an arrow function without `onclick`. Maybe instead of setting `onclick` I should use `this.addEventListener()` for my example. But it's bad style to have `addEventListener` without calling `removeEventListener` when the component's unmounted, right? I know how to do that in React (`componentWillUnmount`), but JS doesn't have destructors so I guess `addEventListener` isn't a good solution here...

Reply:
BTW when you say "instance method" I assume you mean "prototype method". The verbiage can get confusing but from one point of view, an arrow function is arguably an instance method whereas regular methods are not.

Reply:
yes, i mean prototype method. `this.foo = () => {}` isn't what i'd call an instance method, though.

Reply:
I updated the examples and added a new section comparing the different options you mentioned. I think the inheritance section could perhaps still be improved, but for now I changed it to use `onclick` directly and added a note that `SpecialCounter` is not a realistic example.